mlflow/store/artifact/uc_volume_artifact_repo.py

```diff
@@ -0,0 +1,217 @@
+            hostcreds_from_uri = get_databricks_host_creds(databricks_profile_uri)
+            self.get_host_creds = lambda: hostcreds_from_uri
+        else:
+            self.get_host_creds = _get_host_creds_from_default_store()
+
+    def _databricks_api_request(self, endpoint, method, **kwargs):
+        host_creds = self.get_host_creds()
+        return http_request_safe(host_creds=host_creds, endpoint=endpoint, method=method, **kwargs)
+
+    def _volumes_list_api(self, path):
+        host_creds = self.get_host_creds()
+        return http_request(
+            host_creds=host_creds, endpoint=f"{DIR_API_ENDPOINT}{path}", method="GET"
+        )
+
+    def _volumes_download(self, output_path, endpoint):
+        with open(output_path, "wb") as f:
+            response = self._databricks_api_request(endpoint=endpoint, method="GET", stream=True)
+            try:
+                for content in response.iter_content(chunk_size=DOWNLOAD_CHUNK_SIZE):
+                    f.write(content)
+            finally:
+                response.close()
+
+    def _is_directory(self, artifact_path):
+        volume_path = (
+            self._get_volumes_path(artifact_path) if artifact_path else self._get_volumes_path("")
+        )
+        return self._volumes_is_dir(volume_path)
+
+    def _volumes_is_dir(self, volume_path):
+        response = self._databricks_api_request(
+            endpoint=f"{DIR_API_ENDPOINT}{volume_path}", method="HEAD"
+        )
+        try:
+            return response.status_code == 200
+        except KeyError:
+            raise MlflowException(f"Volume path {volume_path} does not exist")
+
+    def _get_volumes_path(self, artifact_path):
+        return "/{}/{}".format(
+            strip_prefix(self.artifact_uri, "dbfs:/"),
+            strip_prefix(artifact_path, "/"),
+        )
+
+    def _get_volumes_endpoint(self, artifact_path):
+        return f"{FILE_API_ENDPOINT}{self._get_volumes_path(artifact_path)}"
+
+    def log_artifact(self, local_file, artifact_path=None):
+        basename = os.path.basename(local_file)
+        if artifact_path:
+            http_endpoint = self._get_volumes_endpoint(posixpath.join(artifact_path, basename))
+        else:
+            http_endpoint = self._get_volumes_endpoint(basename)
+        if os.stat(local_file).st_size == 0:
+            # The API frontend doesn't like it when we post empty files to it using
+            # `requests.request`, potentially due to the bug described in
+            # https://github.com/requests/requests/issues/4215
+            self._databricks_api_request(
+                endpoint=http_endpoint, method="PUT", data="", allow_redirects=False
+            )
+        else:
+            with open(local_file, "rb") as f:
+                self._databricks_api_request(
+                    endpoint=http_endpoint, method="PUT", data=f, allow_redirects=False
+                )
+
+    def log_artifacts(self, local_dir, artifact_path=None):
```

Comment:
question about the retry behavior

Assuming log_artifact is interrupted (e.g. network issue), and part of files have been uploaded successfully, and user want to retry `log_artifact`, then for those files which have been uploaded to UC successfully, will the retry run of ` overwrite the files last time uploaded ? Or error occurs ?

Reply:
overwrite